sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

```diff
@@ -76,41 +98,129 @@ private static void CheckForRedundantConditions(ExplodedGraph explodedGraph, Syn
+        {
+            var unreachableLocations = GetUnreachableLocations(constantNode, constantValue)
+                .ToList();
+
+            var constantText = constantValue.ToString().ToLowerInvariant();
+
+            return unreachableLocations.Count > 0 ?
+                Diagnostic.Create(s2583, constantNode.GetLocation(), messageArgs: constantText, additionalLocations: unreachableLocations) :
+                Diagnostic.Create(s2589, constantNode.GetLocation(), messageArgs: constantText);
+        }
+
+        private static IEnumerable<Location> GetUnreachableLocations(SyntaxNode constantExpression, bool constantValue)
+        {
+            var unreachabExpressions = GetUnreachableExpressions(constantExpression, constantValue)
+                .ToList();
+
+            if (unreachabExpressions.Count > 0)
+            {
+                yield return Location.Create(
+                    constantExpression.SyntaxTree,
+                    GetSpan(unreachabExpressions.First(), unreachabExpressions.Last()));
+            }
+            else
+            {
+                // When the constant expression is the only child of a if/while/etc. statement
+                // it is not reported as unreachable, but regardles we want to check if its
+                // parent has an unreachable branch.
+                var expression = constantExpression as ExpressionSyntax;
+                if (expression != null)
+                {
+                    unreachabExpressions.Add(expression);
+                }
+            }
+
+            var statement = GetUnreachableStatement(unreachabExpressions, constantValue);
+            if (statement != null)
+            {
+                yield return statement.GetLocation();
+            }
+        }
+
+        private static TextSpan GetSpan(SyntaxNode startNode, SyntaxNode endNode)
         {
-            foreach (var alwaysTrue in conditionTrue.Except(conditionFalse))
+            if (startNode.Equals(endNode))
             {
-                context.ReportDiagnostic(Diagnostic.Create(rule, alwaysTrue.GetLocation(), "true"));
+                return startNode.Span;
             }
+            return TextSpan.FromBounds(startNode.SpanStart, endNode.Span.End);
+        }
 
-            foreach (var alwaysFalse in conditionFalse.Except(conditionTrue))
+        private static SyntaxNode GetUnreachableStatement(IEnumerable<ExpressionSyntax> unreachableExpressions, bool constantValue)
+        {
+            var parent = unreachableExpressions
+                .Select(SyntaxHelper.GetSelfOrTopParenthesizedExpression) // unreachable expressions
+                .Select(node => node.Parent is BinaryExpressionSyntax
+                    ? node.Parent.Parent
+                    : node.Parent) // Constant node is the only expression in an if statement condition
+                .FirstOrDefault();
+
+            if (constantValue)

```

Comment:
I would add the following before:

```
if (parent == null)
{
    return null;
}
```
Just to be sure not to try to cast multiple times.